src/cascadia/TerminalApp/ColorScheme.h

```diff
@@ -29,35 +29,76 @@ namespace TerminalAppLocalTests
 
 namespace winrt::TerminalApp::implementation
 {
+// Use this macro to quick implement both the getter and setter for a color property.
+// This should only be used for color types where there's no logic in the
+// getter/setter beyond just accessing/updating the value.
+// This takes advantage of til::color
+#define GETSET_COLORTABLEPROPERTY(name)                                               \
+public:                                                                               \
+    winrt::Windows::UI::Color name() const noexcept { return _table[##name##Index]; } \
+    void name(const winrt::Windows::UI::Color& value) noexcept { _table[##name##Index] = value; }
+
     struct ColorScheme : ColorSchemeT<ColorScheme>
     {
     public:
-        ColorScheme();
+        ColorScheme() = default;
         ColorScheme(hstring name, Windows::UI::Color defaultFg, Windows::UI::Color defaultBg, Windows::UI::Color cursorColor);
-        ~ColorScheme();
 
         void ApplyScheme(const winrt::Microsoft::Terminal::TerminalControl::IControlSettings& terminalSettings) const;
 
         static com_ptr<ColorScheme> FromJson(const Json::Value& json);
         bool ShouldBeLayered(const Json::Value& json) const;
         void LayerJson(const Json::Value& json);
 
-        hstring Name() const noexcept;
-        com_array<Windows::UI::Color> Table() const noexcept;
-        Windows::UI::Color Foreground() const noexcept;
-        Windows::UI::Color Background() const noexcept;
-        Windows::UI::Color SelectionBackground() const noexcept;
-        Windows::UI::Color CursorColor() const noexcept;
+        static Json::Value ToJson(const TerminalApp::ColorScheme& scheme);
+        void UpdateJson(Json::Value& json);
 
         static std::optional<std::wstring> GetNameFromJson(const Json::Value& json);
 
+        com_array<Windows::UI::Color> Table() const noexcept;
+
+        GETSET_PROPERTY(winrt::hstring, Name, L"");
+        GETSET_COLORPROPERTY(Foreground, DEFAULT_FOREGROUND_WITH_ALPHA);
+        GETSET_COLORPROPERTY(Background, DEFAULT_BACKGROUND_WITH_ALPHA);
+        GETSET_COLORPROPERTY(SelectionBackground, DEFAULT_FOREGROUND);
+        GETSET_COLORPROPERTY(CursorColor, DEFAULT_CURSOR_COLOR);
+
+        GETSET_COLORTABLEPROPERTY(Black);
+        GETSET_COLORTABLEPROPERTY(Red);
+        GETSET_COLORTABLEPROPERTY(Green);
+        GETSET_COLORTABLEPROPERTY(Yellow);
+        GETSET_COLORTABLEPROPERTY(Blue);
+        GETSET_COLORTABLEPROPERTY(Purple);
+        GETSET_COLORTABLEPROPERTY(Cyan);
+        GETSET_COLORTABLEPROPERTY(White);
+        GETSET_COLORTABLEPROPERTY(BrightBlack);
+        GETSET_COLORTABLEPROPERTY(BrightRed);
+        GETSET_COLORTABLEPROPERTY(BrightGreen);
+        GETSET_COLORTABLEPROPERTY(BrightYellow);
+        GETSET_COLORTABLEPROPERTY(BrightBlue);
+        GETSET_COLORTABLEPROPERTY(BrightPurple);
+        GETSET_COLORTABLEPROPERTY(BrightCyan);
+        GETSET_COLORTABLEPROPERTY(BrightWhite);

```

Comment:
I almost wonder if we should express these as indices in the settings API. the names from json are only to make it "easier" for people, but... they are meaningless

Reply:
Yeah. The trippy part is connecting this to the Settings UI. The user sees it as "set a value for <color>". But the terminal really sees it as a contiguous table of values. Figured this would be a decent middle ground, but I hate all the `contexpr`s down below for `<color>Index`.

Exposing `<color>Index`, would mean that the Settings UI would look like this:
```c++
void Black_ColorChanged(ColorPicker const&, ColorChangedEventArgs const& event)
{
   _colorScheme.SetColorTable(ColorScheme::BlackIndex, event.Color());
}
```
vs
```c++
void Black_ColorChanged(ColorPicker const&, ColorChangedEventArgs const& event)
{
   _colorScheme.Black(event.Color());
}
```
I don't feel too strongly either way. What are your thoughts?

Reply:
Couldn't it be something more mundane like...

```
void ColorChanged(ColorPicker const&, ColorChangedEventArgs const& event)
{
    auto index = // look up the index attached to this color picker somehow -- via Tag, or some other type of value
    _colorScheme.SetColorTable(index, event.Color());
}
```

?

that way we don't need _16 color changed functions_ and _16 index constants_ ;P

Reply:
Updated. It sounds more and more like the responsibility of converting a "color target" to an index should be on the Settings UI.